designs/node-ownership.md

```diff
@@ -0,0 +1,225 @@
+
+When some other third-party is responsible for registering the node, kubelet only [performs updates on a fixed set of labels](https://github.com/kubernetes/kubernetes/blob/39c76ba2edeadb84a115cc3fbd9204a2177f1c28/pkg/kubelet/kubelet_node_status.go#L215), but does not update other labels which the user might have specified through KubeletConfig. For AWS, this prevents users who want to set dynamic labels through their userData from doing so.
+
+***Relevant GH Issues***
+
+1. [Karpenter ignores extra labels coming from kubelet-extra-args in launchtemplate](https://github.com/aws/karpenter/issues/1681)
+
+## Solutions
+
+Ultimately, the issues listed above involving node creation aren’t exclusive to just creation of the node. The Node is a fundamental component that every Kubernetes user takes some dependency on. As such, it’s reasonable to assume that users have written logic for scheduling/health checks on their nodes that have direct dependency on how a node looks and acts. Fundamentally, it’s a dangerous proposition to mutate the node object in ways that a standard Kubernetes user wouldn’t expect.
+
+As a result, all solutions that solve the node creation problem should:
+
+1. Capture in-flight capacity in a way so that we can quickly complete Karpenter’s provisioning loops and continue to start the next round of provisioning while having a consistent view of cluster state
+2. Completely remove (or, at a minimum, limit) the mutation of the Node so that we don’'t alter the Node in ways that the user doesn't expect
+3. Have little to no impact on cluster performance, including memory usage in etcd and query performance on the kube-apiserver.
+
+Fundamentally, if we do not capture in-flight capacity in the Node object, we need some other place to capture capacity so that we can fulfill #1 listed above. This necessitates some kind of “store” to keep in-flight node data, whether this is an in-memory store or the Kubernetes api-server (etcd) store. The two options presented below highlight both options, including their tradeoffs related to the fundamentals mentioned above.
+
+### Kubernetes CRD Object Store (Machine CRD)
+
+Karpenter will no longer create node objects or launch instances as part of the provisioning loop, but, instead, will create Machine CRs at the completion of the provisioning loop. This machine CR will then be picked up by a separate controller that will launch capacity based on the requirements passed from the provisioning loop and will resolve the static values from the **CreateFleet** response into its status.
+
+#### Impact
+
+1. Introduction of the Machine CR as a representation of a “scheduling decision” to launch by the CloudProvider
+2. Migration of Deprovisioning mechanisms from watching and monitoring the Node object to watching and monitoring the Machine CR
+3. Migration of the Node termination finalizer to Machine to orchestrate the cordoning, draining, and deletion
```

Comment:
Do you really want the finalizer on this object? Just thinking it would be better to handle this async rather than making the delete action actually hang waiting for the instance termination.

Reply:
If I understand this correctly, the first thing that happens is that the Machine CR is created. The actual launch of the instance happens consequently. The Machine resource would then require a random name, unrelated to the instance. If the instance is tagged/labelled with `karpenter.sh/machine-name` we have a direct relationship between the CR and the instance itself.

With this relationship it should be no problem for a "controller" to track instances with no corresponding CR, which would mean the CR has been deleted, and the controller can ensure the instance itself is terminated.

I assume the Machine object's status will also be populated both with the cloud provider's instance ID and the node object once these are available.

Reply:
I think the benefit of having the finalizer on the Machine is that we can migrate the cordon, drain, terminate flow of the Node over to the Machine eventually. In general, we can say Machine Delete results in:

1. Cordon of Node
2. Drain of Node
3. Deletion of Instance
4. Deletion of Node

I could potentially see an argument for handling 3&4 async but if we want to do 1&2, then we require a termination finalizer on the Machine already.

> Just thinking it would be better to handle this async rather than making the delete action actually hang waiting for the instance termination

I'm curious what your concern here is. Are you worried that we may hit failure which could leave the Machine in a hung state due to the finalizer?

Reply:
I am actually more concerned about the client having to wait for the termination to complete before carrying on. If I want to delete a list of Machines, the process would be sequential.

1&2 doesn't require a finalizer. At least I don't quite understand why it would require one.

I actually question if Karpenter should ever cordon & drain, or if this should be up to kubelet (graceful node shutdown) or NTH (which is now also shipped with karpenter). I.e Karpenter should just directly terminate the instance and then rely on other things for 1 & 2.

4. should never be done by Karpenter. It's CCM that deletes Node objects.

Reply:
We want _something_ to take care of load skewing to empty the nodes that Karpenter no longer wants running. We can do that by tainting the node, or via other means. We definitely do want the descheduling to happen.

IMO a finalizer on a Machine is a good way to let Karpenter record (for its own use) an intent to stop using a node. To me, this feels like what finalizers are intended for.

As a response to the scheduled Machine deletion, Karpenter could take descheduling steps such as tainting the node (either `PreferNoSchedule` or `NoSchedule`, or even both at different times) and requesting Pod eviction.

Reply:
Maybe take a step back to consider the ordering @jonathan-innis mentioned:

1. Delete Machine with finalizer
2. Cordon of Node
3. Drain of Node
4. Deletion of Instance
5. Remove finalizer, completing Machine deletion
6. Deletion of Node

As mentioned, 6 is not a Karpenter concern, so I think we can all agree to drop that one.

If we drop the finalizer, the actual condition for instance termination is the same: machine object must be deleted, Node must be cordoned and drained. The intent is the deletion of the machine object and the reconciliation loop picks up this one and acts on it the same way as before. It just doesn't have to remove the finalizer.

Since Karpenter also needs to explicitly track state of EC2 and Machine objects, the finalizer's purpose is _only_ the blocking of deletion.

Reply:
If we didn't have a finalizer on Machine, the problem then becomes that Karpenter loses its record that it owns the node, _before_ cleanup is complete. Much easier to keep the Machine around until the Node actually disappears (or Karpenter gives up).

This also helps with the case where the primary Karpenter is running on a node that we decide to drain, right? If the design makes it easy for one instance of Karpenter to start the scale-in and another to finish the work, that's ideal.

I don't think we need a finalizer on Node.

Reply:
You mean that Karpenter owns the _instance_? The instance ownership should be tracked with a tag in EC2. EC2 has to be the source of truth for that. If a Karpenter pod is started without state, it has to check k8s and EC2 and reconcile. E.g EC2 instance exists, machine object does not means EC2 needs to be drained and terminated.

Reply:
Sort of. How does Karpenter drain a cloud instance when the record of the node-to-instance mapping has already been deleted?

We can't assume that we can use the EC2 API or knowledge about EC2 instance hostnames, because Karpenter has an ambition to support more than one cloud provider.

Reply:
All cloud providers should have an idea of tags or labels on the instance side. So EC2 for AWS, compute engine API on GCP etc. Cloud Controller Manager already does this for Node objects (instance terminated equals Node object removal).

Reply:
Can add that finalizers _may_ be used to track state, but consider it more informational and perhaps use the information to save some cloud API calls. But it can't rely on finalizers for the actual state and Machine cannot be considered the source of truth of whether an instance runs or not.

Reply:
> You mean that Karpenter owns the instance?

Your point is that when Karpenter restarts in the middle of instance deletion/node cordon/drain that we should rediscover it through the tag and re-create the machine? This effectively requires a LIST call on the CloudProvider API to hydrate in any machines that exist that Karpenter isn't tracking. With the finalizer, we don't have to worry about the LIST call since we can be assured that the Machines that exist on the cluster will always be a superset of the instances that exist in EC2

Reply:
> I actually question if Karpenter should ever cordon & drain, or if this should be up to kubelet

I think it's highly important that Karpenter performs this logic to ensure that we don't delete the underlying instance until all the workloads running on that node are removed from that cluster. In the graceful termination case, there is a fixed time for the workload to cleanup, but it won't block indefinitely which is a much riskier proposition to put the customer into.

I could potentially see an argument for forcing deletion of the instance assuming that a user is performing manual deletion of the machine (other cases involve Karpenter automation simulating the termination and not even starting the termination if we know that the drain won't succeed).

Reply:
I'll do this one first since it is simple.

> > I actually question if Karpenter should ever cordon & drain, or if this should be up to kubelet
> 
> I think it's highly important that Karpenter performs this logic to ensure that we don't delete the underlying instance until all the workloads running on that node are removed from that cluster. In the graceful termination case, there is a fixed time for the workload to cleanup, but it won't block indefinitely which is a much riskier proposition to put the customer into.
> 
> I could potentially see an argument for forcing deletion of the instance assuming that a user is performing manual deletion of the machine (other cases involve Karpenter automation simulating the termination and not even starting the termination if we know that the drain won't succeed).

I agree ... I think it is important that the instance is fully drained before termination. Mainly due to the compaction feature and that Karpenter it is not improbable for Karpenter to assume Pods can quickly reschedule when they can't.

Reply:
> > You mean that Karpenter owns the instance?
> 
> Your point is that when Karpenter restarts in the middle of instance deletion/node cordon/drain that we should rediscover it through the tag and re-create the machine? This effectively requires a LIST call on the CloudProvider API to hydrate in any machines that exist that Karpenter isn't tracking. With the finalizer, we don't have to worry about the LIST call since we can be assured that the Machines that exist on the cluster will always be a superset of the instances that exist in EC2

I don't think Machine objects should be recreated (just like Nodes are not recreated). I think the instance should be terminated. I also don't see finalizers as safe enough to guarantee one doesn't leak resources. As mentioned CCM does LIST calls every X seconds against the cloud provider to ensure it's internal list of instances match the cloud provider.

Reply:
In the sketched design from this RFC PR, Karpenter is a controller for Nodes, but only indirectly. It's fine for c-c-m to continue with this behavior, but I'm still expecting that Karpenter's primary reconcile mechanism becomes one where it matches Machines against Nodes and taking appropriate action. Karpenter also triggers indirect Node creations by driving Machines to their desired (ie, running) state.

Kubernetes offers [finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) for cleanup and that's seems to be like a good API mechanism to use here. If we find a Machine with no finalizer set, we can consider adopting it providing it matches the right selector. If someone force-deletes a Machine, we can semi-gracefully terminate the appropriate instance without draining it first (and that person might regret their API call choices).

If Karpenter wants to shut down a server, Karpenter can use its registered finalizer on the Machine to mark the _intent_ to delete and carry out a controlled drain, etc. In my view, that's a really good architectural fit with Kubernetes' API conventions. It also allows any other controller to hook into the same mechanism - for example, a spot interruption, or something like one, could also trigger a delete on the Machine. It's nice to have a design that lets us keep extra complexity apart from Karpenter where we can.